olp-cpp-sdk-dataservice-read/src/VersionedLayerClient.cpp

```diff
@@ -0,0 +1,278 @@
+/*
+ * Copyright (C) 2019 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+#include "olp/dataservice/read/VersionedLayerClient.h"
+
+#include "ApiClientLookup.h"
+#include "generated/api/BlobApi.h"
+#include "generated/api/QueryApi.h"
+#include "generated/model/Api.h"
+#include "olp/core/client/OlpClient.h"
+#include "olp/core/client/OlpClientFactory.h"
+#include "olp/core/context/Context.h"
+#include "olp/core/logging/Log.h"
+#include "olp/core/thread/TaskScheduler.h"
+#include "olp/dataservice/read/Condition.h"
+#include "repositories/ExecuteOrSchedule.inl"
+
+namespace olp {
+namespace dataservice {
+namespace read {
+
+namespace {
+constexpr auto kLogTag = "VersionedLayerClient";
+}
+
+class VersionedLayerClient::Impl final {
+ public:
+  Impl(client::OlpClientSettings client_settings, client::HRN hrn)
+      : client_settings_(std::move(client_settings)), hrn_(std::move(hrn)) {}
+
+  ~Impl() = default;
+
+ private:
+  template <typename Result, typename Error>
+  static bool WaitAndCheck(
+      Condition& condition, const olp::client::CancellationContext& context,
+      const olp::client::ApiResponse<Result, Error>& response,
+      DataResponseCallback callback) {
+    if (!condition.Wait() && !context.IsCancelled()) {
+      callback({{olp::client::ErrorCode::RequestTimeout,
+                 "Network request timed out.", true}});
+      return false;
+    }
+    if (!response.IsSuccessful()) {
+      callback({{olp::client::ErrorCode::ServiceUnavailable,
+                 "Request unsuccessful.", true}});
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * @brief queries Lookup API and returns an OlpClient instance for further
+   * metadata extraction.
+   */
+  static std::tuple<bool, ApiClientLookup::ApiClientResponse>
+  LookupQueryServiceStep(olp::client::CancellationContext& context,
+                         const std::shared_ptr<client::OlpClient>& olp_client,
+                         const olp::client::HRN& hrn,
+                         DataResponseCallback callback) {
+    OLP_SDK_LOG_DEBUG_F(kLogTag, "step 1 - LookupQueryServiceStep");
```

Comment:
```suggestion
    OLP_SDK_LOG_TRACE(kLogTag, "step 1 - LookupQueryServiceStep");
```
I think debug is too much for those logs. Maybe even remove those logs.